doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,129 @@
+- When they run the `COPY` command, to keep things simple we'll block the UI till the operation
+is complete, similar to any other query.
+- If successful, the output of the `COPY` command should return the number of rows and
+the number of bytes/files written to S3.
+- If the operation fails midway for some reason or is cancelled by the user, we should show
+a notice to the users that incomplete data could have been written to the S3 path and will
+need to be cleaned up.
+- Later we can provide an option to the user to always append a unique query ID to the path,
+so they can just run the same query again and again without needing to clean up previous data.
+
+## Implementation
+
+### Using Compute cluster
+Currently existing supported `COPY ... TO STDOUT` command fetches the result to `environmentd`.
+This will not be advisable for `COPY ... TO S3` as the data size could be large. Therefore,
+we should trigger a dataflow in a `clusterd` which should eventually write to S3 using
+user's aws connection details. The next question being, should it be a Storage cluser or a
+Compute cluster (at least till the time the cluster unification is done).
+
+Compute already has a concept of ["Compute Sinks"](https://github.com/MaterializeInc/materialize/blob/v0.77.1/src/compute-types/src/sinks.rs#L100-L103), which include Subscribe
+and Persist (for materialized views). We can create another Compute S3 sink, which will
+actually read from persist and write to S3. Keeping this in a compute cluster also means,
+- We eventually will support `SELECT` queries in a `COPY ... TO S3` command and that can
+be a natural extension if things are already in compute.
+- Via the Persist Compute Sink, there's already precedence to talking to S3, we should
+be able to leverage some existing code.
+- This also will be consistent with the behaviour that a user can select the compute cluster
+where they want their query to run like they do for other queries.
+
+#### Other Alternative: Storage cluster
+The other alternative would have been to actually make use of a Storage cluster. If we think
+of `COPY ... TO S3` as a one time sink, then it might also make sense to have it in the Storage
+cluster. Especially if the initial scope does not include `SELECT` queries which
+would have required spinning up compute dataflows.
+Reasons to not do so,
+- As mentioned above, we will eventually support `SELECT` queries in `COPY ... TO S3`. The
+only reason it would have made sense to do this in storage would be to do as a quick stop gap
+if the effort is drastically less. It does not appear to be so.
+- Unlike Compute, Storage does not have a concept of running temporary/one-off dataflows.Everything which runs, like sources or sinks, do so continuously once scheduled.
+
+### Output file format
+We should start with a single file format and add more if there's more user interest. Currently,
+there has been request for parquet, so that will be the initial supported format.
```

Comment:
If we go with the compute implementation, I'm not sure that it will be more work to support arbitrary queries! In fact, it might actually be harder to _disallow_ arbitrary queries.

Reply:
I had tried some changes locally to see how things will fit. It seemed like it would be fairly easy to only support tables/mvs? Basically we'll have to do something similar to https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/optimize/subscribe.rs#L164-L194 (and not do the other match arm which is for a query).

Reply:
I updated to start with csv format first cc @hlburak